resource/aws_dx_private_virtual_interface: Update resource to support MTU param

slapula opened the pull request:
Fixes #6136

Changes proposed in this pull request:

* Updated `aws_dx_private_virtual_interface` resource to support new MTU parameter

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAwsDxPrivateVirtualInterface_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAwsDxPrivateVirtualInterface_basic -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAwsDxPrivateVirtualInterface_basic
--- SKIP: TestAccAwsDxPrivateVirtualInterface_basic (0.00s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	0.029s
```

ewbankkit commented:
Related:
* https://github.com/terraform-providers/terraform-provider-aws/pull/6142
* https://github.com/terraform-providers/terraform-provider-aws/pull/6143

ewbankkit commented:
Please add verification of the `mtu` and `jumbo_frame_enabled` attributes to the acceptance tests; something like:

```go
resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "mtu", "9001"),
resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "jumbo_frame_enabled", "1"),
```

ewbankkit commented on aws/resource_aws_dx_private_virtual_interface.go:
```diff
@@ -84,6 +84,16 @@ func resourceAwsDxPrivateVirtualInterface() *schema.Resource {
+				Computed: true,
+				ForceNew: true,
+			},
+			"jumbo_frame_enabled": {
```
Nitpick - In the [documentation](https://docs.aws.amazon.com/directconnect/latest/UserGuide/set-jumbo-frames-vif.html) the feature is called _Jumbo Frame**s**_. For consistency I suggest naming the attribute `jumbo_frames_enabled` or `jumbo_frame_capable` (to match the API).

slapula replied:
Oops, I meant to have this match the API but I just spaced on it.  This will be corrected in my next commit.

ewbankkit commented on aws/resource_aws_dx_private_virtual_interface.go:
```diff
@@ -84,6 +84,16 @@ func resourceAwsDxPrivateVirtualInterface() *schema.Resource {
 				Computed: true,
 				ForceNew: true,
 			},
+			"mtu": {
```
See `ValidateFunc` comment: https://github.com/terraform-providers/terraform-provider-aws/pull/6142/files#r224971730.

ewbankkit commented on aws/resource_aws_dx_private_virtual_interface.go:
```diff
@@ -84,6 +84,16 @@ func resourceAwsDxPrivateVirtualInterface() *schema.Resource {
+				Type:     schema.TypeInt,
+				Default:  1500,
+				Optional: true,
+				ForceNew: true,
```
See `Update` comment: https://github.com/terraform-providers/terraform-provider-aws/pull/6142/files#r224971890.

slapula commented:
@ewbankkit I believe I have addressed your comments with this latest commit.  I ran out of time today to test and write tests so I will likely circle back to that tomorrow or Tuesday.  It looks to me that my changes "should work" so if you have spare cycles to test before I get to it please feel free.

slapula commented on aws/resource_aws_dx_private_virtual_interface.go:
```diff
@@ -219,3 +235,22 @@ func dxPrivateVirtualInterfaceWaitUntilAvailable(d *schema.ResourceData, conn *d
 		})
 }
+
+func IntInSlice(valid []int) schema.SchemaValidateFunc {
```
Again, haven't fully tested this yet but would this be a good candidate to move to the core validation library?

ewbankkit replied:
Yes, I think it would but given the churn in the codebase with v0.12 it is probably best to wait until that settles down.

ewbankkit replied:
Probably better to move this to [`validators.go`](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/validators.go) and rename it to something like `validateIntInSlice`.

ewbankkit commented:
OK, I'll look at testing this week.

ewbankkit commented:
@slapula I made a couple of mainly cosmetic changes and pushed to my `slapula-dx-private-virtual-interface-mtu` branch:
* According to the [documentation](https://docs.aws.amazon.com/directconnect/latest/UserGuide/set-jumbo-frames-vif.html), after an update is made to the virtual interface its state will be `pending` until the update is complete so I added an `Update` value to the schema's `Timeouts` and added a call to `dxPrivateVirtualInterfaceWaitUntilAvailable` after calling the update API (this changed the signature to a function shared by all the VIF attributes so there was some cascading call site changes)
* Simplified your acceptance tests - No need to do the MTU update in `TestAccAwsDxPrivateVirtualInterface_basic`, just do it in `TestAccAwsDxPrivateVirtualInterface_mtuUpdate`
* Minor documentation tweaks

I should be able to run the acceptance tests in the next couple of days - I'll make any modifications on that branch and you can then merge back.

ewbankkit commented:
@slapula I took your latest code (fixed a merge error) and ran acceptance tests for the new test case.
With one minor change

```go
resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "jumbo_frame_capable", "true")
->
resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "jumbo_frame_capable", "1")
```
the test passes:

```console
$ DX_CONNECTION_ID=dxcon-xxxxxxxx AWS_DEFAULT_REGION=us-east-1 make testacc TESTARGS='-run=TestAccAwsDxPrivateVirtualInterface_mtuUpdate'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAwsDxPrivateVirtualInterface_mtuUpdate -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAwsDxPrivateVirtualInterface_mtuUpdate
--- PASS: TestAccAwsDxPrivateVirtualInterface_mtuUpdate (1252.51s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1270.641s
```

If you merge the latest changes from https://github.com/ewbankkit/terraform-provider-aws/tree/slapula-dx-private-virtual-interface-mtu I can the run all the `aws_dx_private_virtual_interface` acceptance tests on the complete code.

slapula commented:
@ewbankkit Thanks for the help on this!  I totally spaced on the `Update` and `WaitUntilAvailable` calls so good catch. Anyway, I have your changes pulled in and Travis is green.  If we're good here, I can start porting these changes over to #6142.

ewbankkit commented on aws/resource_aws_dx_private_virtual_interface_test.go:
```diff
@@ -79,6 +79,49 @@ func TestAccAwsDxPrivateVirtualInterface_dxGateway(t *testing.T) {
+					testAccCheckAwsDxPrivateVirtualInterfaceExists("aws_dx_private_virtual_interface.foo"),
+					resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "name", vifName),
+					resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "mtu", "1500"),
+					resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "jumbo_frame_capable", "1"),
```
Need to change this to
```go
resource.TestCheckResourceAttr("aws_dx_private_virtual_interface.foo", "jumbo_frame_capable", "true"),
```

or else will get errors like

```console
$ DX_CONNECTION_ID=dxcon-xxxxxxxx AWS_DEFAULT_REGION=us-east-1 make testacc TESTARGS='-run=TestAccAwsDxPrivateVirtualInterface_mtuUpdate'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAwsDxPrivateVirtualInterface_mtuUpdate -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAwsDxPrivateVirtualInterface_mtuUpdate
--- FAIL: TestAccAwsDxPrivateVirtualInterface_mtuUpdate (667.21s)
    testing.go:538: Step 0 error: Check failed: Check 4/4 error: aws_dx_private_virtual_interface.foo: Attribute 'jumbo_frame_capable' expected "1", got "true"
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	668.068s
make: *** [testacc] Error 1
```

ewbankkit commented:
@slapula Left a comment on changing `1` to `true` in the new acceptance test, once that's committed  I can run the full acceptance tests. Yes, makes sense to merge the changes into the other PR.

slapula commented:
Oh ok, I misread your previous comment. I'll update that in a moment.

ewbankkit commented:
OK, tested the code (serially as there is a problem with running the acceptance tests in parallel, https://github.com/terraform-providers/terraform-provider-aws/issues/6244, which shouldn't block this PR).

```console
$ git rev-parse HEAD
da69517c8d918611b8ab9599b7e82ee1445c8be4
$ DX_CONNECTION_ID=dxcon-xxxxxxxx AWS_DEFAULT_REGION=us-east-1 make testacc TESTARGS='-run=TestAccAwsDxPrivateVirtualInterface_mtuUpdate'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAwsDxPrivateVirtualInterface_mtuUpdate -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAwsDxPrivateVirtualInterface_mtuUpdate
--- PASS: TestAccAwsDxPrivateVirtualInterface_mtuUpdate (1252.51s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	(cached)
$ DX_CONNECTION_ID=dxcon-xxxxxxxx AWS_DEFAULT_REGION=us-east-1 make testacc TESTARGS='-run=TestAccAwsDxPrivateVirtualInterface_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAwsDxPrivateVirtualInterface_basic -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAwsDxPrivateVirtualInterface_basic
=== PAUSE TestAccAwsDxPrivateVirtualInterface_basic
=== CONT  TestAccAwsDxPrivateVirtualInterface_basic
--- PASS: TestAccAwsDxPrivateVirtualInterface_basic (409.98s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	410.875s
$ DX_CONNECTION_ID=dxcon-xxxxxxxx AWS_DEFAULT_REGION=us-east-1 make testacc TESTARGS='-run=TestAccAwsDxPrivateVirtualInterface_dxGateway'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAwsDxPrivateVirtualInterface_dxGateway -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAwsDxPrivateVirtualInterface_dxGateway
=== PAUSE TestAccAwsDxPrivateVirtualInterface_dxGateway
=== CONT  TestAccAwsDxPrivateVirtualInterface_dxGateway
--- PASS: TestAccAwsDxPrivateVirtualInterface_dxGateway (589.50s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	590.384s
```

ewbankkit commented:
This PR is now ready for review.

bflad commented:
This has been released in [version 1.42.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1420-october-31-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
